### PR TITLE
Add instruct tuning support to LoRA training

### LINKED ISSUE #484: Instruct tuning for lora/finetune?
Please correct me if I'm wrong, but it looks like the current examples for lora training all build a loss function around completion, which lines up with the lora example of using only the 'text' field from the jsonl dataset.

Are there any forks or plans to allow for instruct tuning, where the input is an input prompt, and the loss function is targeting the input/output pair?

Or did I miss something?

Thanks!

### PR DESCRIPTION
**Pull request by anupamme**
Fixes #484

Add support for instruct tuning with input/output pairs and alternative loss functions.

* **llms/mlx_lm/lora.py**
  - Add `CompletionsDataset` class to support input/output pairs.
  - Modify `Dataset` class to handle different dataset types.
  - Update `main` function to include new dataset type.

* **llms/mlx_lm/tuner/trainer.py**
  - Modify `default_loss` function to support alternative loss functions.
  - Add new `instruct_loss` function for instruct tuning.

* **llms/mlx_lm/LORA.md**
  - Add instructions for instruct tuning with input/output pairs.
  - Update documentation to include alternative loss functions.

* **llms/tests/test_datasets.py**
  - Add tests for `CompletionsDataset` and `create_dataset` functions.

* **llms/tests/test_trainer.py**
  - Add tests for `default_loss` and `instruct_loss` functions.

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/ml-explore/mlx-examples/pull/1211?shareId=c9b0c4c8-f90c-4258-8177-0d6942de8524).

### CODE REVIEW

**Comment by awni** [comment]
Thanks for your contribution, but I think this was mostly closed by #1103 which adds a `--mask-prompt`  flag to enable fine-tuning only on the completion.

I'll close this in favor of that since it landed. If there is additional functionality here, please send a new PR rebased on main.
